### DISCUSSION TITLE
Running autocoders in generate step

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#2220] [Ideas]
Would it be possible to run the autocoders in the `generate`, rather than `build` step? I very often find myself purging my build directory, but then have to build my deployment - several minutes - in order to be able to find inherited methods in my editor.

**Comment by LeStarch** [comment]
It would be possible.  It does move time into the `generate` step.  Typically, we try to avoid this because `generate` is a single-threaded execution whereas a build can be run in parallel with the `-j x` flag. Thus adding time to the non-parallel step adds time to the critical path.

That being said, another option could be to add a `FPRIME_CACHE_AUTOCODERS` flag that turns such an option on.  This would allow users to select which they prefer.

**Comment by SterlingPeet** [comment]
There is a technical problem thats not well aligned with the normal cmake development workflow expectations.  Normally, the user would expect to be able to generate once and then build many times, hence the focus on speedup in that area.  The leads the developer to expect:

- cmake generate will take a medium amount of time
- initial build takes a long amount of time
- subsequent re-builds are _very_ short (because only the changed code is compiled)

Unfortunately the autocoded targets are [or at least used to be] very difficult to detect rebuild needs.  Thus, changing things that require them to be rebuild leads to broken builds, which is highly confusing.  That ultimately leads to regenerating the project, an undesireable workflow, but the currently recommended workflow for F Prime projects.

Personally, if there was a manual flag that could be set to re-build all the autocoded stuff, that would be a suitable workaround to minimize the re-compilation and re-generation requirements, but it does mean that the original request is not fully satisfied.  You would still need to build once to get the autocoded headers importable by your IDE.

**Reply by JackNWhite** [reply]
When you say "build once", do you mean build everything once, or build once with the flag, so that only the autocoders run? The latter would be a definite improvement.

**Reply by SterlingPeet** [reply]
The latter: effectively a flag that tells the autocoder targets they need to re-run on this specific build invocation
